File: android/app/src/main/java/com/noop/data/StandardHrMapping.kt
package com.noop.data

import com.noop.protocol.StandardHrContact
import org.json.JSONObject

/** Durable mapping for standard-BLE contact readings; legacy HR rows have no companion event. */
object StandardHrMapping {
    const val CONTACT_EVENT_KIND = "STANDARD_HR_CONTACT"

    fun contactEvent(ts: Long, contact: StandardHrContact): EventEntry = EventEntry(
        ts = ts,
        kind = CONTACT_EVENT_KIND,
        payloadJSON = "{\"contact\":\"${contact.storageValue}\"}",
    )

    /**
     * Parse a persisted [CONTACT_EVENT_KIND] [EventRow.payloadJSON]. Throws [org.json.JSONException]
     * (or [IllegalArgumentException] for an unknown value) so a parse failure is not the same as
     * "no contact event" — absence is an empty query, not a skipped row.
     */
    fun contactSample(row: EventRow): StandardHrContactSample {
        val raw = JSONObject(row.payloadJSON).getString("contact")
        val contact = StandardHrContact.fromStorageValue(raw)
            ?: throw IllegalArgumentException("unknown STANDARD_HR_CONTACT value: $raw")
        return StandardHrContactSample(row.ts, contact)
    }
}

data class StandardHrContactSample(val ts: Long, val contact: StandardHrContact)

File: Packages/WhoopStore/Sources/WhoopStore/StandardHRMapping.swift
import Foundation
import WhoopProtocol

/// Pure, testable mapping from a single standard-BLE Heart-Rate reading (0x2A37) onto the
/// datastore's `Streams` shape, so an isolated generic-strap source (`StandardHRSource` in the
/// app target) can persist its samples through the SAME `StreamStore.insert` path the WHOOP
/// pipeline uses — without duplicating the row-construction logic in the app target where it
/// can't be unit-tested.
///
/// A chest strap (Polar / Wahoo / Coospo / Garmin HRM / Amazfit Helio broadcast) only ever
/// reports HR, (optionally) R-R intervals, and its standard contact flags over 0x2A37; every other
/// stream (spo2, skin temp, resp, gravity, steps, ppgHr, battery) is left empty.
public enum StandardHRMapping {
    /// `event.kind` used for a standard BLE sensor-contact reading. The generic event table is the
    /// established durable stream for additive decoded signals, so this needs no schema migration.
    public static let contactEventKind = "STANDARD_HR_CONTACT"

    /// Build a `Streams` carrying one HR sample and zero-or-more R-R intervals, all stamped at the
    /// same wall-clock `ts` (unix seconds). Pure → unit-testable.
    public static func samples(fromHR hr: Int, rr: [Int], contact: StandardHRContact? = nil,
                               at ts: Int) -> Streams {
        let events = contact.map { [
            WhoopEvent(ts: ts, kind: contactEventKind, payload: ["contact": .string($0.rawValue)])
        ] } ?? []
        return Streams(
            hr: [HRSample(ts: ts, bpm: hr)],
            rr: rr.map { RRInterval(ts: ts, rrMs: $0) },
            events: events
        )
    }
}

/// One persisted standard-BLE sensor-contact reading. Legacy HR rows have no companion event and are
/// therefore absent from this stream instead of being guessed as any contact state.
public struct StandardHRContactSample: Equatable, Sendable {
    public let ts: Int
    public let contact: StandardHRContact
    public init(ts: Int, contact: StandardHRContact) {
        self.ts = ts
        self.contact = contact
    }
}
